hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
-                    if self.marked.contains(&node_id) {
-                        return;
-                    }
-                    // Ignore handoff nodes.
-                    if matches!(self.nodes[node_id], Node::Handoff) {
-                        return;
-                    }
+        (node_subgraph, subgraph_nodes)
+    };
 
-                    for &(next_back, _) in self.preds[node_id].values() {
-                        if self.node_union.same_set(node_id, next_back) {
-                            self.visit(next_back);
-                        }
-                    }
+    (new_preds, new_succs, node_subgraph, subgraph_nodes)
+}
 
-                    let repr_node = self.node_union.find(node_id);
-                    if !self.subgraph_nodes.contains_key(repr_node) {
-                        self.subgraph_nodes.insert(repr_node, Default::default());
+fn find_subgraph_strata(
+    nodes: &mut SlotMap<GraphNodeId, Node>,
+    new_preds: &mut AdjList,
+    new_succs: &mut AdjList,
+    node_subgraph: &mut SecondaryMap<GraphNodeId, GraphSubgraphId>,
+    subgraph_nodes: &mut SlotMap<GraphSubgraphId, Vec<GraphNodeId>>,
+    barrier_crossers: &[(GraphNodeId, GraphNodeId, IndexInt, InputBarrier)],
+) -> Result<SecondaryMap<GraphSubgraphId, usize>, ()> {
+    // Determine subgraphs's stratum number.
+    // Find SCCs ignoring negative edges, then do TopoSort on the resulting DAG.
+
+    // Generate subgraph graph.
+    let mut subgraph_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+    let mut subgraph_succs: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+
+    for (node_id, node) in nodes.iter() {
+        if matches!(node, Node::Handoff) {
+            for &(pred, _) in new_preds[node_id].values() {
+                let pred_sg = node_subgraph[pred];
+                for &(succ, succ_idx) in new_succs[node_id].values() {
+                    if barrier_crossers.contains(&(pred, succ, succ_idx, InputBarrier::Epoch)) {
+                        continue;
                     }
-                    self.subgraph_nodes[repr_node].push(node_id);
-                    self.marked.insert(node_id);
+                    let succ_sg = node_subgraph[succ];
+                    subgraph_preds.entry(succ_sg).or_default().push(pred_sg);
+                    subgraph_succs.entry(pred_sg).or_default().push(succ_sg);
                 }
             }
+        }
+    }
 
-            let (marked, subgraph_nodes) = Default::default();
-            let mut sg_topo_sort = SubgraphTopoSort {
-                nodes: &flat_graph.nodes,
-                preds: &new_preds,
-                node_union: &mut node_union,
-                marked,
-                subgraph_nodes,
-            };
-            for node_id in flat_graph.nodes.keys() {
-                sg_topo_sort.visit(node_id);
+    let scc = {
+        // https://en.wikipedia.org/wiki/Kosaraju%27s_algorithm
+        fn visit(
```

Comment:
comment the graph traversal and output order

Reply:
this should be its own method in `graph_algorithms.rs`